### PR TITLE
Preview IoT Provisioning Client: Certificate Signing Request, TrustBundle

### PR DESCRIPTION
**Pull request by CIPop**
Preview features:
- Azure IoT Provisioning Client Certificate Signing Request
- Azure IoT Provisioning Client TrustBundle
- Azure IoT Provisioning Client Receiving Custom Allocation Payloads

Includes unit tests, sample and documentation.

**Note:** the current *preview* implementation performs in-place JSON un-escaping which will alter the original MQTT receive buffer. A design discussion is tracked by https://github.com/Azure/azure-sdk-for-c/issues/2173.

**Deprecation notice:** `az_iot_provisioning_client_get_request_payload()` is deprecated in favor of `az_iot_provisioning_client_register_get_request_payload()`. The new API _requires_ an initialized `az_iot_provisioning_client_payload_options` structure (using `az_iot_provisioning_client_payload_options_default()`).

### CODE REVIEW

File: sdk/tests/iot/provisioning/test_az_iot_provisioning_client.h [hunk 1]
```diff
@@ -1,9 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
-// Placeholder for int test_iot_provisioning_(); declarations
-
 int test_az_iot_provisioning_client();
 int test_az_iot_provisioning_client_sas_token();
 int test_az_iot_provisioning_client_parser();
-int test_az_iot_provisioning_client_payload();
+int test_az_iot_proviosining_client_register_get_request_payload();
```
Comment (danewalton) [inline sdk/tests/iot/provisioning/test_az_iot_provisioning_client.h:7]:
```suggestion
int test_az_iot_provisioning_client_register_get_request_payload();
```

**Review by danewalton — APPROVED** [review]

File: sdk/docs/iot/resources/csr.puml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+Device -> Device: Creates Certificate Signing Request using **operational** key-pair
+Device -> DPS: Authenticate using SAS, TPM or Client X509 **onboarding** credentials
+DPS -> Device: ACK
+Device -> DPS: register (registrationId, Body: **operational** CSR)
+DPS -> Device: ACK
+DPS -> CertificateAuthority: **operational** CSR
+DPS <- CertificateAuthority: **operational** X509 Certificate
+Device -> DPS: query [...]
+DPS -> Device: hubFQDN, deviceId, **operational** X509 Certificate
+Device -> HUB: Authenticate using **operational** X509 Certificate
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/docs/iot/resources/csr.puml:12]: DPS disconnect is not required but it's something that most devices would likely do to use less resources.

File: sdk/docs/iot/using_certificate_signing_requests.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+```
+
+2. Generate a PKCS#10 Certificate Signing Request (CSR) from the keypair:
+
+Below, replace `my-device-id` with the device ID.
+
+_Important:_ The device ID *must* match the one configured for the registration ID. Devices are not allowed to submit requests with Subject Certificate Name other than the configured device ID.
+
+```
+openssl req -new -key key.pem -out request.csr -subj '/CN=my-device-id'
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/docs/iot/using_certificate_signing_requests.md:65]: No, this won't work with the registration ID. It is the device ID, as noted in the *Important* section above.

File: sdk/inc/azure/iot/az_iot_provisioning_client.h [hunk 4]
```diff
@@ -17,6 +17,7 @@
 #ifndef _az_IOT_PROVISIONING_CLIENT_H
 #define _az_IOT_PROVISIONING_CLIENT_H
 
+#include <azure/core/az_json.h>
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/inc/azure/iot/az_iot_provisioning_client.h:20]: Correct: left-over from when we wanted json readers. Will remove.

File: sdk/inc/azure/iot/az_iot_provisioning_client.h [hunk 5]
```diff
@@ -262,6 +263,23 @@ typedef struct
...
   az_span error_timestamp;
+
+  /**
+   * An optional custom payload received from the service.
+   */
+  az_span payload;
+
+  /**
+   * An optional set of X509 Certification Authorities received from the service.
+   * Review.
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/inc/azure/iot/az_iot_provisioning_client.h:274]: Goot catch! Removing.

File: sdk/inc/azure/iot/az_iot_provisioning_client.h [hunk 6]
```diff
@@ -426,19 +445,66 @@ AZ_NODISCARD az_result az_iot_provisioning_client_query_status_get_publish_topic
...
+ * @pre \p mqtt_payload_size must be greater than 0.
+ * @pre \p out_mqtt_payload_length must not be `NULL`.
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK The payload was created successfully.
+ * @retval #AZ_ERROR_NOT_ENOUGH_SPACE The buffer is too small.
+ */
+AZ_NODISCARD az_result az_iot_provisioning_client_register_get_request_payload(
+    az_iot_provisioning_client const* client,
+    az_span custom_payload_property,
+    az_iot_provisioning_client_payload_options const* options,
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/inc/azure/iot/az_iot_provisioning_client.h:500]: I've left it the same signature as the deprecated API so I've assumed it's in the right position.
+ @ahsonkhan @RickWinter as FYI

Maybe we move all parameters in the options (i.e. remove `custom_payload_property`)?

File: sdk/samples/iot/README.md [hunk 7]
```diff
@@ -652,6 +652,12 @@ This section provides an overview of the different samples available to run and
 
   This [sample](https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/samples/iot/paho_iot_provisioning_sas_sample.c) registers a device with the Azure IoT Device Provisioning Service. It will wait to receive the registration status before disconnecting. SAS authentication is used.
 
+### _Preview_ IoT Provisioning CSR and TrustBundle Sample
+
+- *Executable:* `paho_iot_provisioning_csr_trustbundle_sample`
+
+  This [sample](https://github.com/Azure/azure-sdk-for-c/blob/feature/iot-certificate-management/sdk/samples/iot/paho_iot_provisioning_csr_trustbundle_sample.c) registers a device with the Azure IoT Device Provisioning Service. X509 onboarding authentication (X509 in this case) is used to submit a *certificate signing request* (CSR) for an operational certificate. The sample shows usage for *custom allocation payload*, *Trust Bundle* and obtaining the *operational X509 client certificate*.
+
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/samples/iot/README.md:660]: Good point: Final documentation needs to be added to docs.microsoft.com. Meanwhile, I'll add links to the docs/iot documents I've added.

File: sdk/src/azure/iot/az_iot_provisioning_client.c [hunk 8]
```diff
@@ -276,6 +281,35 @@ AZ_INLINE az_result _az_iot_provisioning_client_parse_payload_error_code(
   return AZ_ERROR_ITEM_NOT_FOUND;
 }
 
+AZ_INLINE az_result
+_az_iot_provisioning_client_get_json_object_span(az_json_reader* jr, az_span* out_span)
+{
+  *out_span = jr->token.slice;
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/src/azure/iot/az_iot_provisioning_client.c:287]: It's not obvious but it is. I've used the pattern of storing things in variables already allocated on the stack but I could use a local instead.
Line 308 does a diff between the original slice and the slice after skipping all children JSON elements.

File: sdk/tests/iot/provisioning/test_az_iot_provisioning_client.h (hunk 1 above)
Comment (CIPop) [reply sdk/tests/iot/provisioning/test_az_iot_provisioning_client.h:7]: Will replace-all.

File: sdk/docs/iot/resources/csr.puml (hunk 2 above)
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/docs/iot/resources/csr.puml]: Updated the PUML. In the future we'll need to switch to Mermaid to make changes faster.

**Pull request merged**